In [1]:
import pandas as pd
import numpy as np
import datetime
import math

In [2]:
comments_df = pd.read_csv('../data/comments_df.csv')

In [3]:
comments_df.head(15)

Unnamed: 0       id   parent_id             username         time_created  \
0            0  14m8mf4         NaN            Tungsten_  2023-06-29 10:54:44   
1            1  jq5du0z  t3_14m8mf4            Tungsten_  2023-06-30 11:33:11   
2            2  jq0dgzx  t3_14m8mf4       ProudBlackMatt  2023-06-29 11:16:15   
3            3  jq0cg7k  t3_14m8mf4         TomatoCanned  2023-06-29 11:09:47   
4            4  jq0f52k  t3_14m8mf4         bad-fengshui  2023-06-29 11:26:41   
5            5  jq0mhrx  t3_14m8mf4      Pancake_muncher  2023-06-29 12:12:39   
6            6  jq0mlbs  t3_14m8mf4              suberry  2023-06-29 12:13:16   
7            7  jq0cvvn  t3_14m8mf4  Puzzled-Painter3301  2023-06-29 11:12:31   
8            8  jq0jtzk  t3_14m8mf4             e9967780  2023-06-29 11:55:52   
9            9  jq0kfzq  t3_14m8mf4                  NaN  2023-06-29 11:59:41   
10          10  jq0iyws  t3_14m8mf4       OkartoIceCream  2023-06-29 11:50:34   
11          11  jq20w8p  t3_14m8mf4        ShalomHasaeyo  2023-06-29 17:36:41   
12          12  jq0b8ts  t3_14m8mf4         SteadfastEnd  2023-06-29 11:02:09   
13          13  jq2fp75  t3_14m8mf4           MsNewKicks  2023-06-29 19:26:21   
14          15  jq0dxui  t3_14m8mf4              pal2002  2023-06-29 11:19:14   

                                                flair  \
0                                                 NaN   
1                                                 NaN   
2                                    Chinese-American   
3                                                 NaN   
4                                                 NaN   
5                                                 NaN   
6                                                 NaN   
7                                                 NaN   
8                                                 NaN   
9                                                 NaN   
10                                                NaN   
11                                                NaN   
12                                                NaN   
13  First Of Her Name, Queen ABG, 나쁜 기집애, Blocker ...   
14                                                NaN   

                                                 body  \
0   [Megathread] Supreme Court Ruling on Affirmati...   
1   Thanks to everyone who engaged in insightful a...   
2   I would prefer using a process that takes into...   
3   u/Tungsten_, Thanks for creating a section jus...   
4   As with anything related to Asians in politics...   
5   Yet colleges will allow alumni and doners in e...   
6   I just hated Affirmative Action as a distracti...   
7   My own feeling is that I was never in love wit...   
8   Anti Asian racism whether against East Asians ...   
9   Can we overturn legacy and athlete admissions ...   
10  I want to remind people that in California, on...   
11  I think this is great news for our nation. Whi...   
12  My main concern is that, since the Court did n...   
13  There is obviously a lot to it and no real eas...   
14  I’m sure selective schools like Harvard will f...   

                                           tokens_new  \
0   ['Megathread', 'Supreme', 'Court', 'Ruling', '...   
1   ['Thanks', 'to', 'everyone', 'who', 'engaged',...   
2   ['I', 'would', 'prefer', 'using', 'a', 'proces...   
3   ['u/Tungsten_,', 'Thanks', 'for', 'creating', ...   
4   ['As', 'with', 'anything', 'related', 'to', 'A...   
5   ['Yet', 'colleges', 'will', 'allow', 'alumni',...   
6   ['I', 'just', 'hated', 'Affirmative', 'Action'...   
7   ['My', 'own', 'feeling', 'is', 'that', 'I', 'w...   
8   ['Anti', 'Asian', 'racism', 'whether', 'agains...   
9   ['Can', 'we', 'overturn', 'legacy', 'and', 'at...   
10  ['I', 'want', 'to', 'remind', 'people', 'that'...   
11  ['I', 'think', 'this', 'is', 'great', 'news', ...   
12  ['My', 'main', 'concern', 'is', 'that', 'since...   
13  ['There', 'is', 'obviously', 'a', 'lot', 'to',...   
14  ['I', '’m', 'sure', 's

## Divide comments and posts created pre-SCOTUS and post-SCOTUS ruling

In [4]:
def str_to_datetime(str_):
    if isinstance(str_, str):
        return datetime.datetime.fromisoformat(str_)
    else:
        return None
        

comments_df['time_created_datetime'] = comments_df['time_created'].apply(lambda x: str_to_datetime(x))

#### Impute rows without dates

In [5]:
no_datetime = comments_df[pd.isna(comments_df['time_created'])]

print(no_datetime.shape)

(0, 12)


In [6]:
ruling_date = datetime.datetime(2023, 6, 29)
print(ruling_date)

2023-06-29 00:00:00


In [7]:
first_row = comments_df.iloc[0]
first_time_str = first_row['time_created']

print(type(first_time_str))

first_datetime = datetime.datetime.fromisoformat(first_time_str)
print(type(first_datetime))

print(isinstance(first_time_str, str))

<class 'str'>
<class 'datetime.datetime'>
True


In [11]:
comments_df.head(20)

Unnamed: 0       id   parent_id             username         time_created  \
0            0  14m8mf4         NaN            Tungsten_  2023-06-29 10:54:44   
1            1  jq5du0z  t3_14m8mf4            Tungsten_  2023-06-30 11:33:11   
2            2  jq0dgzx  t3_14m8mf4       ProudBlackMatt  2023-06-29 11:16:15   
3            3  jq0cg7k  t3_14m8mf4         TomatoCanned  2023-06-29 11:09:47   
4            4  jq0f52k  t3_14m8mf4         bad-fengshui  2023-06-29 11:26:41   
5            5  jq0mhrx  t3_14m8mf4      Pancake_muncher  2023-06-29 12:12:39   
6            6  jq0mlbs  t3_14m8mf4              suberry  2023-06-29 12:13:16   
7            7  jq0cvvn  t3_14m8mf4  Puzzled-Painter3301  2023-06-29 11:12:31   
8            8  jq0jtzk  t3_14m8mf4             e9967780  2023-06-29 11:55:52   
9            9  jq0kfzq  t3_14m8mf4                  NaN  2023-06-29 11:59:41   
10          10  jq0iyws  t3_14m8mf4       OkartoIceCream  2023-06-29 11:50:34   
11          11  jq20w8p  t3_14m8mf4        ShalomHasaeyo  2023-06-29 17:36:41   
12          12  jq0b8ts  t3_14m8mf4         SteadfastEnd  2023-06-29 11:02:09   
13          13  jq2fp75  t3_14m8mf4           MsNewKicks  2023-06-29 19:26:21   
14          15  jq0dxui  t3_14m8mf4              pal2002  2023-06-29 11:19:14   
15          16  jq0w387  t3_14m8mf4         Ok-Value5827  2023-06-29 13:12:38   
16          17  jq22me5  t3_14m8mf4              j3ychen  2023-06-29 17:49:07   
17          18  jq28kbb  t3_14m8mf4     HappyPineapple11  2023-06-29 18:32:41   
18          19  jq0heyu  t3_14m8mf4           alandizzle  2023-06-29 11:40:55   
19          20  jq28fcf  t3_14m8mf4          gamesrgreat  2023-06-29 18:31:37   

                                                flair  \
0                                                 NaN   
1                                                 NaN   
2                                    Chinese-American   
3                                                 NaN   
4                                                 NaN   
5                                                 NaN   
6                                                 NaN   
7                                                 NaN   
8                                                 NaN   
9                                                 NaN   
10                                                NaN   
11                                                NaN   
12                                                NaN   
13  First Of Her Name, Queen ABG, 나쁜 기집애, Blocker ...   
14                                                NaN   
15                                                NaN   
16                                          Taiwanese   
17                                                NaN   
18                                     I'm Asian. Hi.   
19                                                NaN   

                                                 body  \
0   [Megathread] Supreme Court Ruling on Affirmati...   
1   Thanks to everyone who engaged in insightful a...   
2   I would prefer using a process that takes into...   
3   u/Tungsten_, Thanks for creating a section jus...   
4   As with anything related to Asians in politics...   
5   Yet colleges will allow alumni and doners in e...   
6   I just hated Affirmative Action as a distracti...   
7   My own feeling is that I was never in love wit...   
8   Anti Asian racism whether against East Asians ...   
9   Can we overturn legacy and athlete admissions ...   
10  I want to remind people that in California, on...   
11  I think this is great news for our nation. Whi...   
12  My main concern is that, since the Court did n...   
13  There is obviously a lot to it and no real eas...   
14  I’m sure selective schools like Harvard will f...   
15  TBH, I don't know the depth of how affirmative...   
16  People can continue to debate the legal implic...   
17  The only issue with affirmative action was tha...   
18  You know what I h

In [8]:
first_datetime = comments_df.iloc[0]['time_created_datetime']
print(type(first_datetime))

<class 'pandas._libs.tslibs.timestamps.Timestamp'>


In [9]:
pre_scotus = comments_df[comments_df['time_created_datetime'] < datetime.datetime(2023, 6, 29)]

In [10]:
pre_scotus

Unnamed: 0       id   parent_id          username         time_created  \
337          362   hqic5b         NaN  JesusSinfulHands  2020-07-13 11:22:51   
338          363  fxy6xpv   t3_hqic5b       DJIisStupid  2020-07-13 12:38:51   
339          364  fxy3l3x   t3_hqic5b               NaN  2020-07-13 12:12:24   
340          365  fxxxtp3   t3_hqic5b  JesusSinfulHands  2020-07-13 11:26:06   
341          366  fxzzm8j   t3_hqic5b               NaN  2020-07-13 22:00:26   
...          ...      ...         ...               ...                  ...   
3388        3729  epcqlfa  t1_elt4c17         arientyse  2019-05-29 14:15:14   
3389        3730   4ziaf0         NaN             unkle  2016-08-25 07:44:54   
3390        3731  d6witvi   t3_4ziaf0          Swordbow  2016-08-25 14:47:22   
3391        3732  d6z9lhb   t3_4ziaf0           IRVCath  2016-08-27 17:02:04   
3392        3733  d6w9qin   t3_4ziaf0   BletchTheWalrus  2016-08-25 11:40:47   

                       flair  \
337                      NaN   
338                      NaN   
339                      NaN   
340                      NaN   
341                      NaN   
...                      ...   
3388                     NaN   
3389               senyorito   
3390                     NaN   
3391  Fil-Am, 1.5 Generation   
3392                     NaN   

                                                   body  \
337   A big gap in support for affirmative action ha...   
338   wonder how this is split between income levels...   
339   Very interesting that the split for affirmativ...   
340   Taken from this [story](http://aapidata.com/bl...   
341   Frankly as a ABC currently in college, I have ...   
...                                                 ...   
3388  Please don't refer to us as "the blacks." It's...   
3389  From Self-Interest to Collective Morality: How...   
3390  The writer takes good nuggets but integrates t...   
3391  Interestingly enough, even though Filipinos ar...   
3392                                               Nope   

                                             tokens_new  \
337   ['A', 'big', 'gap', 'in', 'support', 'for', 'a...   
338   ['wonder', 'how', 'this', 'is', 'split', 'betw...   
339   ['Very', 'interesting', 'that', 'the', 'split'...   
340   ['Taken', 'from', 'this', 'story](http://aapid...   
341   ['Frankly', 'as', 'a', 'ABC', 'currently', 'in...   
...                                                 ...   
3388  ['Please', 'do', "n't", 'refer', 'to', 'us', '...   
3389  ['From', 'Self', 'Interest', 'to', 'Collective...   
3390  ['The', 'writer', 'takes', 'good', 'nuggets', ...   
3391  ['Interestingly', 'enough', 'even', 'though', ...   
3392                                           ['Nope']   

                                      normalized_tokens  \
337   ['big', 'gap', 'support', 'affirmative', 'acti...   
338   ['wonder', 'split', 'income', 'level', 'migrat...   
339   ['interesting', 'split', 'affirmative', 'actio...   
340   ['take', 'story](http://aapidata.com', 'blog',...   
341   ['frankly', 'abc', 'currently', 'college', 're...   
...                                                 ...   
3388  ['refer', 'black', 'okay', 'black', 'people', ...   
3389  ['self', 'interest', 'collective', 'morality',...   
3390  ['writer', 'take', 'good', 'nugget', 'integrat...   
3391  ['interestingly', 'filipino', 'relatively', 'd...   
3392                                           ['nope']   

      normalized_tokens_count  word_count time_created_datetime  
337                        10          17   2020-07-13 11:22:51  
338                         6          11   2020-07-13 12:38:51  
339                        82         171   2020-07-13 12:12:24  
340                        81         137   2020-07-13 11:26:06  
341                        54          98   2020-07-13 22:00:26  
...                       ...         ...                   ...  
3388                        6          20   2019-05-29 14:15:14  
3389 

In [11]:
post_scotus = comments_df[comments_df['time_created_datetime'] >= datetime.datetime(2023, 6, 29)]

In [12]:
post_scotus

Unnamed: 0       id   parent_id              username  \
0              0  14m8mf4         NaN             Tungsten_   
1              1  jq5du0z  t3_14m8mf4             Tungsten_   
2              2  jq0dgzx  t3_14m8mf4        ProudBlackMatt   
3              3  jq0cg7k  t3_14m8mf4          TomatoCanned   
4              4  jq0f52k  t3_14m8mf4          bad-fengshui   
...          ...      ...         ...                   ...   
334          358  jq24mg4  t1_jq1itxi    Different-Rip-2787   
335          359  jq3gway  t1_jq3f4ro                   NaN   
336          361  jq2ectl  t1_jq25ucy    Different-Rip-2787   
3368        3708  1belws6         NaN   DudeGuyBroPalBuddie   
3369        3709  kz983nz  t3_1belws6  Extension_River_9901   

             time_created             flair  \
0     2023-06-29 10:54:44               NaN   
1     2023-06-30 11:33:11               NaN   
2     2023-06-29 11:16:15  Chinese-American   
3     2023-06-29 11:09:47               NaN   
4     2023-06-29 11:26:41               NaN   
...                   ...               ...   
334   2023-06-29 18:03:45               NaN   
335   2023-06-30 00:57:44               NaN   
336   2023-06-29 19:16:06               NaN   
3368  2024-03-14 09:02:59               NaN   
3369  2024-04-12 11:37:00               NaN   

                                                   body  \
0     [Megathread] Supreme Court Ruling on Affirmati...   
1     Thanks to everyone who engaged in insightful a...   
2     I would prefer using a process that takes into...   
3     u/Tungsten_, Thanks for creating a section jus...   
4     As with anything related to Asians in politics...   
...                                                 ...   
334   I fail to see how this is 'detrimental' to any...   
335   >We have test scores to look at\n\nIf you beli...   
336   > Those same options were also available to As...   
3368  How White Supremacy uses Taiwan This is a long...   
3369  White supremacy wants less Asian people in the...   

                                             tokens_new  \
0     ['Megathread', 'Supreme', 'Court', 'Ruling', '...   
1     ['Thanks', 'to', 'everyone', 'who', 'engaged',...   
2     ['I', 'would', 'prefer', 'using', 'a', 'proces...   
3     ['u/Tungsten_,', 'Thanks', 'for', 'creating', ...   
4     ['As', 'with', 'anything', 'related', 'to', 'A...   
...                                                 ...   
334   ['I', 'fail', 'to', 'see', 'how', 'this', 'is'...   
335   ['>', 'We', 'have', 'test', 'scores', 'to', 'l...   
336   ['>', 'Those', 'same', 'options', 'were', 'als...   
3368  ['How', 'White', 'Supremacy', 'uses', 'Taiwan'...   
3369  ['White', 'supremacy', 'wants', 'less', 'Asian...   

                                      normalized_tokens  \
0     ['megathread', 'supreme', 'court', 'ruling', '...   
1     ['thank', 'engage', 'insightful', 'respectful'...   
2     ['prefer', 'process', 'take', 'account', 'pove...   
3     ['u/tungsten_,', 'thank', 'create', 'section',...   
4     ['relate', 'asians', 'politic', 'm', 'see', 'l...   
...                                                 ...   
334   ['fail', 'detrimental', 'black', 'brown', 'stu...   
335   ['test', 'score', 'look', 'believe', 'well', '...   
336   ['option', 'available', 'asam', 'student', 'di...   
3368  ['white', 'supremacy', 'use', 'taiwan', 'long'...   
3369  ['white', 'supremacy', 'want', 'asian', 'peopl...   

      normalized_tokens_count  word_count time_created_datetime  
0                          62          78   2023-06-29 10:54:44  
1                           9          20   2023-06-30 11:33:11  
2                          52         103   2023-06-29 11:16:15  
3                         126         269   2023-06-29 11:09:47  
4                          25          59   2023-06-29 11:26:41  
...                       ...         ...                   ...  
334                        37          88   2023-06-29 18:03:45  
335                        45

(OLD) Breakdown of datetimes:
- Before SCOTUS decision: 2554 rows
- After SCOTUS decision: 304 rows

- time_created is NaN: 534 rows

Since we're missing a significant number of rows with a date, collect row's date from earlier post/comment

(UPDATED) Breakdown of datetimes:
- Before SCOTUS decision: 3054 rows
- After SCOTUS decision: 339 rows

- time_created is NaN: 0 rows

#### Look at the earliest June 29 posts/comments to see if they were truly created after SCOTUS decision was annoucned